python/hail/matrixtable.py

```diff
@@ -2878,14 +2889,16 @@ def _select_rows(self, caller, key_struct=None, value_struct=None, pk_size=None)
         analyze(caller, row, self._row_indices, {self._col_axis})
         base, cleanup = self._process_joins(row)
 
-        return cleanup(MatrixTable(base._jvds.selectRows(row._ast.to_hql(), new_key)))
+        return cleanup(MatrixTable(base._jvds.selectRows(str(row._ir), new_key)))
 
-    @typecheck_method(key_struct=nullable(expr_struct()),
-                      value_struct=nullable(expr_struct()),
-                      pk_size=nullable(int))
-    def _select_rows_processed(self, key_struct=None, value_struct=None, pk_size=None):
-        row, new_key = self._make_row(key_struct, value_struct, pk_size)
-        return MatrixTable(self._jvds.selectRows(row._ast.to_hql(), new_key))
+    @typecheck_method(key_struct=expr_struct())
+    def _select_rows_processed(self, key_struct):
+        new_key = (list(key_struct.keys()), [])
+        keys = Env.get_uid()
+        k_ref = Ref(keys, key_struct.dtype)
+        fields = [(n, GetField(k_ref, n)) for (n, t) in key_struct.dtype.items()]
+        row_ir = Let(keys, key_struct._ir, InsertFields(self.row._ir, fields))
```

Comment:
why do we need a let here? It's only used once inside

Reply:
I'm actually just very confused in general about this, and could use some help understanding

Reply:
`fields` uses `key_struct` as many times as there are fields in the key struct

Reply:
(as for why it got rewritten to generate IR directly---`self._make_row` calls `StructExpression.annotate(**fields)`, which calls `unify_all` on the field expressions. When I was rewriting `annotate` I made it unify the original struct expression, also, which I didn't realize wasn't the original behavior at first, so it complained because the source of the original struct is different from the source of the annotations in this case. I did this instead of changing it back, because I think the expr unification should include the struct, but I could do that instead and deal with this separately.)